### PR TITLE
Fix Units-of-Measurement (UoM) handling

### LINKED ISSUE #1107: Temperature setpoint sends update in incorrect unit
My temperature setpoint is in Fahrenheit units. When adjusting the setpoint, the selected number is interpreted as Celsius, resulting in the completely wrong temperature. For example, if I set 72 as the setpoint, the app sends 72 Celsius to the server, which is 161.6 Fahrenheit, which is what then shows in the app. The app properly displays the current and setpoint temperatures in Fahrenheit units. The issue is just with adjusting the setpoint. This issue does not occur in the web interface.

### PR DESCRIPTION
**Pull request by maniac103**
Parse units from item states, and add them back when updating state.

Fixes #1107

### CODE REVIEW

**Comment by maniac103** [comment]
Did you check the JSON server response for the sitemap? I guess state is transmitted in °C, so not sure what to do there?

**Comment by mueller-ma** [comment]
The first widget uses `%unit%` in its label, the other have hardcoded units which are also shown as label, e.g. `"label": "F[45.0 °F]"` has Fahrenheit hardcoded.
````
  "widgets": [
    {
      "widgetId": "000100",
      "type": "Setpoint",
      "label": "default [7.2 °C]",
      "icon": "number",
      "mappings": [],
      "step": 1.2,
      "item": {
        "link": "http://openhab/rest/items/Temperature",
        "state": "7.2000003 °C",
        "stateDescription": {
          "pattern": "%.1f %unit%",
          "readOnly": false,
          "options": []
        },
        "editable": false,
        "type": "Number:Temperature",
        "name": "Temperature",
        "label": "Outside Temperature",
        "tags": [],
        "groupNames": []
      },
      "widgets": []
    },
    {
      "widgetId": "000101",
      "type": "Setpoint",
      "label": "C[7.2 °C]",
      "icon": "number",
      "mappings": [],
      "item": {
        "link": "http://openhab/rest/items/Temperature",
        "state": "7.2000003 °C",
        "stateDescription": {
          "pattern": "%.1f %unit%",
          "readOnly": false,
          "options": []
        },
        "editable": false,
        "type": "Number:Temperature",
        "name": "Temperature",
        "label": "Outside Temperature",
        "tags": [],
        "groupNames": []
      },
      "widgets": []
    },
    {
      "widgetId": "000102",
      "type": "Setpoint",
      "label": "F[45.0 °F]",
      "icon": "number",
      "mappings": [],
      "state": "44.96000054 °F",
      "item": {
        "link": "http://openhab/rest/items/Temperature",
        "state": "7.2000003 °C",
        "stateDescription": {
          "pattern": "%.1f %unit%",
          "readOnly": false,
          "options": []
        },
        "editable": false,
        "type": "Number:Temperature",
        "name": "Temperature",
        "label": "Outside Temperature",
        "tags": [],
        "groupNames": []
      },
      "widgets": []
    },
    {
      "widgetId": "000103",
      "type": "Setpoint",
      "label": "K[280.4 K]",
      "icon": "number",
      "mappings": [],
      "state": "280.3500003 K",
      "item": {
        "link": "http://openhab/rest/items/Temperature",
        "state": "7.2000003 °C",
        "stateDescription": {
          "pattern": "%.1f %unit%",
          "readOnly": false,
          "options": []
        },
        "editable": false,
        "type": "Number:Temperature",
        "name": "Temperature",
        "label": "Outside Temperature",
        "tags": [],
        "groupNames": []
      },
      "widgets": []
    }
  ]

````

**Comment by maniac103** [comment]
Ah, widget state vs. item state. I'll see what we can do there.

**Comment by maniac103** [comment]
@mueller-ma I tried to follow your item + widgets example, but I don't get `%unit%` in the format string:

```
                    {
                        "widgetId": "0103",
                        "type": "Setpoint",
                        "label": "Temperature K [275,3 K]",
                        "icon": "number",
                        "mappings": [

                        ],
                        "minValue": 0,
                        "maxValue": 40,
                        "step": 0.5,
                        "state": "275.3311777215700867 K",
                        "item": {
                            "link": "http://192.168.2.26:8080/rest/items/Temperature",
                            "state": "2.1811777215700867 °C",
                            "stateDescription": {
                                "pattern": "%.1f °C",
                                "readOnly": false,
                                "options": [

                                ]
                            },
                            "editable": false,
                            "type": "Number:Temperature",
                            "name": "Temperature",
                            "tags": [

                            ],
                            "groupNames": [

                            ]
                        },
                        "widgets": [

                        ]
                    }
 ```

Is there anything special I need to do there?

File: mobile/src/main/java/org/openhab/habdroid/model/Widget.java [hunk 1]
```diff
@@ -140,9 +143,14 @@ public Widget build() {
             maxValue(Math.max(minValue(), maxValue()));
             step(Math.abs(step()));
 
+            if (state() == null && item() != null) {
+                state(item().state());
+            }
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/model/Widget.java]: Can remove this, handled by determineWidgetState already. (Need to pass item state into state() for OH1)

**Comment by mueller-ma** [comment]
My items and sitemap file:
````
Number:Temperature Temperature  "Temperature[%.1f %unit%]"
````

````
Setpoint item=Temperature label="default [%.1f %unit%]" minValue=-100 maxValue=100
Setpoint item=Temperature label="C[%.1f °C]" minValue=-100 maxValue=100
Setpoint item=Temperature label="F[%.1f °F]" minValue=-1000 maxValue=1000
Setpoint item=Temperature label="K[%.1f K]" minValue=-100 maxValue=1000
````

With your latest change everything seems to work now.

**Comment by maniac103** [comment]
It's probably the item label, which you explicitly specify while I left it at defaults (like in your example in https://github.com/openhab/openhab-android/pull/1146#pullrequestreview-185712735).

Sounds like we're good then?

**Comment by mueller-ma** [comment]
Yes, but the unit tests fail

**Comment by maniac103** [comment]
It'll need a few days until I can actually test the variable slider support: its introduction is pretty new (https://github.com/eclipse/smarthome/pull/6777) and is not in the smarthome snapshot used by the default Eclipse configuration yet, and I'm not sure how to make it use the source version instead of the jar version.

**Comment by maniac103** [comment]
Tested the variable slider support, seems fine.

**Comment by mueller-ma** [comment]
I tested this again and when I change the Kelvin setpoint to 400K, the item is changed to 400°C.

**Comment by maniac103** [comment]
Will check this. If you have the build still at hand, can you check in the logs what's sent?

**Comment by maniac103** [comment]
@mueller-ma Fixed. The item type wasn't parsed correctly and thus 'Number:Temperature' became type enum value 'None', which triggered the logic in https://github.com/openhab/openhab-android/pull/1146/commits/784302d202277da9320b9906527bf7fc89bea936

**Comment by mueller-ma** [comment]
Using `Slider item=Temperature label="default [%.1f %unit%]"` crashes the app. The setpoints are working now.

**Comment by maniac103** [comment]
> Using Slider item=Temperature label="default [%.1f %unit%]" crashes the app.

Fixed (even though using a slider with %unit% doesn't seem meaningful, because the slider range changes depending on what unit was used last time when setting the state).

File: mobile/src/main/java/org/openhab/habdroid/model/ParsedState.java [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+@AutoValue
+public abstract class ParsedState implements Parcelable {
+    public static class NumberState implements Parcelable {
+        public final Number mValue;
```
[the remark this answers is not in the document]
Comment (maniac103) [reply mobile/src/main/java/org/openhab/habdroid/model/ParsedState.java]: Actually, this PR should already fix it as it uses the formatting specifier in the stateDescription for the setpoint values ... can you confirm it does? ;-)

[the remark this answers is not in the document]
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/model/ParsedState.java]: Now the setpoints always open for me with the minValue displayed and not the current one.

[the remark this answers is not in the document]
Comment (mueller-ma) [reply mobile/src/main/java/org/openhab/habdroid/model/ParsedState.java]: Oops, I haven't used the latest version of your PR. Setpoints are working and #1148 is fixed.

File: mobile/src/main/java/org/openhab/habdroid/model/ParsedState.java [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+
+import com.google.auto.value.AutoValue;
+
+import java.util.IllegalFormatConversionException;
+import java.util.Locale;
+import java.util.UnknownFormatConversionException;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import androidx.annotation.NonNull;
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/model/ParsedState.java]: This seems to be a valid checkstyle issue.

File: mobile/src/main/java/org/openhab/habdroid/util/Util.java [hunk 4]
```diff
@@ -205,12 +221,13 @@ public static void sendItemCommand(AsyncHttpClient client, String itemUrl, Strin
                 new AsyncHttpClient.StringResponseHandler() {
             @Override
             public void onFailure(Request request, int statusCode, Throwable error) {
-                Log.e(TAG, "Got command error " + error.getMessage());
+                Log.e(TAG, "Sending command '" + command + "' to " + itemUrl +
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/util/Util.java]: This as well

**Comment by maniac103** [comment]
Should be good now.

**Comment by mueller-ma** [comment]
This PR introduced two LGTM alerts: https://lgtm.com/projects/g/openhab/openhab-android/alerts/?mode=list

**Comment by maniac103** [comment]
I'm impressed, both are legit. The error is what caused the section switch woes and is fixed in #1185; the warning should be fixed by comparing against `number`, not `state`.

**Pull request merged**